BQ queries fail on GitLab due to reported lack of reservation

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/57746

```
2022-11-29 23:03:31,773 WARNING ThreadPoolExecutor-0_13: Connection pool is full, discarding connection: jade.datarepo-dev.broadinstitute.org. Connection pool size: 1
2022-11-29 23:03:31,774 DEBUG   ThreadPoolExecutor-0_13: _request(…) -> 1.788s 200 b'{"id":"3011a1cf-1388-4819-8091-fee55e878ea1","name":"hca_dev_d3446f0c30f34a12b7c36af877c7bb2d__20210901_20210903","description":"Create snapshot hca_dev_d3446f0c30f34a12b7c36af877c7bb2d__20210901_20210903","createdDate":"2021-09-03T22:02:34.888433Z","co...'
Traceback (most recent call last):
  File "/builds/ucsc/azul/scripts/post_deploy_tdr.py", line 86, in <module>
    main()
  File "/builds/ucsc/azul/scripts/post_deploy_tdr.py", line 81, in main
    verify_sources()
  File "/builds/ucsc/azul/scripts/post_deploy_tdr.py", line 54, in verify_sources
    raise e
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/builds/ucsc/azul/src/azul/terra.py", line 490, in check_bigquery_access
    self.run_sql(f'''
  File "/builds/ucsc/azul/src/azul/terra.py", line 519, in run_sql
    result = job.result()
  File "/build/.venv/lib/python3.9/site-packages/google/cloud/bigquery/job/query.py", line 1499, in result
    do_get_result()
  File "/build/.venv/lib/python3.9/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func
    return retry_target(
  File "/build/.venv/lib/python3.9/site-packages/google/api_core/retry.py", line 190, in retry_target
    return target()
  File "/build/.venv/lib/python3.9/site-packages/google/cloud/bigquery/job/query.py", line 1489, in do_get_result
    super(QueryJob, self).result(retry=retry, timeout=timeout)
  File "/build/.venv/lib/python3.9/site-packages/google/cloud/bigquery/job/base.py", line 728, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/build/.venv/lib/python3.9/site-packages/google/api_core/future/polling.py", line 137, in result
    raise self._exception
google.api_core.exceptions.BadRequest: 400 Cannot run query: project does not have the reservation in the data region
Location: us-central1
Job ID: d68f4718-d8e7-4ed3-9106-4c4d0de41de3
make[1]: *** [Makefile:100: post_deploy] Error 1
make[1]: Leaving directory '/builds/ucsc/azul'
make: *** [Makefile:95: auto_deploy] Error 2
```

----

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/59161

```
ERROR: test_can_bundle_configured_catalogs (integration_test.CanBundleScriptIntegrationTest) (catalog='dcp2-it', repository=Config.Catalog.Plugin(name='tdr_hca'))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 1472, in test_can_bundle_configured_catalogs
    self._test_catalog(catalog)
…
  File "/build/.venv/lib/python3.9/site-packages/google/cloud/bigquery/job/base.py", line 728, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/build/.venv/lib/python3.9/site-packages/google/api_core/future/polling.py", line 137, in result
    raise self._exception
google.api_core.exceptions.BadRequest: 400 Cannot run query: project does not have the reservation in the data region
Location: us-central1
Job ID: daf22823-63f2-471a-bccc-9f39f5695cf5
----------------------------------------------------------------------
Ran 13 tests in 542.729s
FAILED (errors=1, skipped=3)
make: *** [Makefile:238: integration_test] Error 1
```

--- a/src/azul/terra.py
+++ b/src/azul/terra.py
@@ -24,6 +24,7 @@ from furl import (
     furl,
 )
 from google.api_core.exceptions import (
+    BadRequest,
     Forbidden,
     InternalServerError,
     ServiceUnavailable,
@@ -517,10 +518,13 @@ class TDRClient(SAMClient):
                 job: QueryJob = bigquery.query(query)
                 try:
                     result = job.result()
-                except (Forbidden, InternalServerError, ServiceUnavailable) as e:
+                except (BadRequest, Forbidden, InternalServerError, ServiceUnavailable) as e:
                     if delay is None:
                         raise e
                     elif isinstance(e, Forbidden) and 'Exceeded rate limits' not in e.message:
+                        raise e
+                    elif (isinstance(e, BadRequest)
+                          and 'project does not have the reservation in the data region' not in e.message):
                         raise e
                     else:
                         log.warning('BigQuery job error during attempt %i/%i. Retrying in %is.',

